---
title: "Assignment 1"
author: "ChengYue Zhang"
output:
  html_document:
    toc: yes
    toc_depth: '2'
    df_print: paged
  html_notebook:
    toc: yes
    toc_depth: 2
bibliography: A1.bib
---

# Download Packages
* R (@r)
* [BiocManager](https://CRAN.R-project.org/package=BiocManager) (@morgan2021a)
* [GEOmetadb](https://bioconductor.org/packages/release/bioc/html/GEOmetadb.html) (@mccarthy2012a)
* [edgeR](https://bioconductor.org/packages/release/bioc/html/edgeR.html) (@robinson2010a, @mccarthy2012a, @chen2016a)
* [limma](https://bioconductor.org/packages/release/bioc/html/limma.html) (@ritchie2015a)
* [biomaRt](https://bioconductor.org/packages/release/bioc/html/biomaRt.html) (@durinck2009a, @durinck2005a)
* [GEOquery](https://bioconductor.org/packages/release/bioc/html/GEOquery.html) (@davis2007a)
* [Biobase & BioGenerics](https://bioconductor.org/packages/release/bioc/html/Biobase.html) (@huber2015a)
```{r Install package, message=FALSE, warning=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE)){
  install.packages("BiocManager")}

if (!requireNamespace("GEOmetadb", quietly = TRUE)){
  BiocManager::install("GEOmetadb")}

if (!requireNamespace("ggplot2", quietly = TRUE)){
  install.packages("ggplot2")}

if (!requireNamespace("knitr", quietly = TRUE)){
  install.packages("knitr")}

if (!requireNamespace("RSQLite", quietly = TRUE)){
  install.packages("RSQLite")}

if (!requireNamespace("edgeR", quietly = TRUE)){
  BiocManager::install("edgeR")}

if (!requireNamespace("limma", quietly = TRUE)){
  BiocManager::install("limma")}

if (!requireNamespace("biomaRt", quietly = TRUE)){
  BiocManager::install("biomaRt")}

if (!requireNamespace("GEOquery", quietly = TRUE)){
  BiocManager::install("GEOquery")}

if (!requireNamespace("Biobase", quietly = TRUE)){
  BiocManager::install("Biobase")}


if (!requireNamespace("odbc", quietly = TRUE)){
  install.packages("odbc")}

if (!requireNamespace("dplyr", quietly = TRUE)){
  install.packages("dplyr")}

if (!requireNamespace("DBI", quietly = TRUE)){
  install.packages("DBI")}

if (!requireNamespace("kableExtra", quietly = TRUE)){
  install.packages("kableExtra")}

```

Load packages
```{r Load library, message=FALSE}
library("GEOmetadb")
library("ggplot2")
library("knitr")
library("RSQLite")
library("edgeR")
library("limma")
library("biomaRt")
library("GEOquery")
library("Biobase")
library("odbc")
library("dplyr")
library("DBI")
library("kableExtra")
```

# Select an Expression Data Set

## 1. Download the data

Get the GEO description of my dataset GSE162515
```{r Get dataset, message=FALSE}
gse <- getGEO("GSE162515",GSEMatrix=FALSE)
kable(data.frame(head(Meta(gse))), format = "html")
```
<br>
Information about platform
```{r Information about platform, message=FALSE}
current_gpl <- names(GPLList(gse))[1]
current_gpl_info <- Meta(getGEO(current_gpl))

current_gpl_info$title

current_gpl_info$last_update_date

current_gpl_info$organism

```
<br>

Get the expression data
```{r Get expression data, message=FALSE}
sfiles <- getGEOSuppFiles('GSE162515')
```
<br>

There are four supplement files. Here, I will choose [4] which is the RNAseq raw counts file.
```{r List supplement files, echo = TRUE, message = TRUE}
fnames <- rownames(sfiles)
ntm_exp <- read.delim(fnames[4],header=TRUE,
                check.names = FALSE)
fnames
```
<br>

Information about expression data (i.e. gene names, chromosome, start/end position, sequence length) are from column 88 to 92 in the table. Move these to the first column for better visualization and processing.
```{r Info about expression data}
expInfos <- ntm_exp[, seq(88, 92)]

ntm_exp <- cbind(expInfos, ntm_exp[, seq(-88, -92)])
```
<br>
Check the table
```{r Check expression data}
kable(ntm_exp[1:15,1:15], format = "html")
```



## 2. Assess data quality

How many genes do I have measurements for?
```{r Dimension of loaded data}
dim(ntm_exp)
```
<br>
Are any of genes duplicated? and Why?

```{r Look for duplicated genes step 1}
summarized_gene_counts <- sort(table(ntm_exp$Gene),
                               decreasing = TRUE)
```
<br>
Only output those with frequency greater than 1
```{r Obtain high freq rows}
kable(summarized_gene_counts[which(summarized_gene_counts>1)[1:10]], format = 'html')
```
<br>

So yes, there are duplicated genes, but not very significant. The highest frequency of duplicate is only 2. These might be the PCR duplicates that result from sequencing two or more copies of the exact same cDNA fragment. I do not neet to filter them out.

Right now, the column names for the count data are very informative, as shown below:
```{r List column names}
colnames(ntm_exp)[1:15]
```
### Use the GSE series matrix to make the column names informative

Obtain the series matrix file through GEOquery.
```{r Obtain series matrix, message=FALSE}
gseFile <- getGEO('GSE162515',GSEMatrix=TRUE)
gseInfo <- pData(phenoData(gseFile[[1]]))

```
<br>

Filter out the samples that are not RNA-seq.
```{r Filter out non-RNA-seq data}
rnaSeq <- gseInfo[grepl("RNA", gseInfo$title), c(1, 6, 8)]
rnaSeq[1:10,]
```
<br>

Process sample names
```{r Process sample names, results='hide'}
samples <- lapply(rnaSeq$title, FUN=function(x){unlist(strsplit(x,
                        split = " "))[2]})
sample_types <- substr(samples,1,nchar(samples)-1)
newColNames <- c(colnames(ntm_exp)[1:5], sample_types)
colnames(ntm_exp) <- newColNames
```
<br>
**Group the data**

There are three groups of samples, where "T" indicates "Thyroid tumor", "N" indicates "Normal thyroid, and "M" indicates "Cervical LymphNode Metastases". 
```{r Group data}
#get the 2 and third token from the column names
samples <- data.frame(lapply(colnames(ntm_exp)[6:92], FUN=function(x){unlist(strsplit(x,
                        split = "\\-"))[c(2,3)]}))
colnames(samples) <- colnames(ntm_exp)[6:92]
rownames(samples) <- c("individual","tissue_type")
samples <- data.frame(t(samples))
```
<br>
From the paper (@sanghi2021a), the authors indicates that there are 27 "N" type samples, 30 "T" type samples and 30 "M" type samples. Check whether I have obtained all of them.
```{r Samples in each group}
sum(grepl("N", samples$tissue_type))
# [1] 27
sum(grepl("T", samples$tissue_type))
# [1] 30
sum(grepl("M", samples$tissue_type))
# [1] 30
```


# Filter out low counts

```{r Filter out low counts}
#translate out counts into counts per million using 
#the edgeR package function cpm
cpms <- edgeR::cpm(ntm_exp[,6:92])
rownames(cpms) <- ntm_exp$Gene
# get rid of low counts
keep <- rowSums(cpms >1) >=3
ntm_exp_filtered <- ntm_exp[keep,]
```


<br>
**What does that do to the dataset?**

The size of our dataset gets smaller.
```{r Difference before and after filtering on the expression data}
dim(ntm_exp)
dim(ntm_exp_filtered)
```
<br>
**What does that do to the duplicates?**

The duplicates do not change. The highest frequency of duplicate is still 2.
```{r Obtain high freq}
summarized_gene_counts_filtered <- sort(table(ntm_exp$Gene),
                               decreasing = TRUE)

kable(summarized_gene_counts_filtered[
  which(summarized_gene_counts_filtered>1)[1:10]],
  format="html")
```



# Normalization
## Distribution of our data
### Boxplot

The following is the box plot of the original count data, before filtering out any low counts. We can see that there is some fluctuation in the medians, and the overall median is close to 0, indicating the existence of low-expressed or non-expressed genes.
```{r Boxplot non-filtered, warning=FALSE,message=FALSE}
data1plot <- log2(cpm(ntm_exp[,6:92]))
boxplot(data1plot, xlab = "Samples", ylab = "log2 CPM",
        las = 2, cex = 0.5, cex.lab = 0.5,
        cex.axis = 0.5, main = "Non-filtered RNASeq Samples")
#draw the median on each box plot
abline(h = median(apply(data1plot, 2, median)),
       col = "green", lwd = 0.6, lty = "dashed")
```




After filtering out the low count genes, the new boxplot is shown below. The central tendency of the count data is close to log2CPM of 5, indicating a success remove of low count genes, and the genes in the samples have relatively high expression level. Interestingly, the data is already well-aligned, so we would expect the normalization step improves minimally on the quality of the dataset.
```{r Boxplot filtered, warning=FALSE,message=FALSE}
data2plot <- log2(cpm(ntm_exp_filtered[,6:92]))
boxplot(data2plot, xlab = "Samples", ylab = "log2 CPM",
        las = 2, cex = 0.5, cex.lab = 0.5,
        cex.axis = 0.5, main = "Filtered RNASeq Samples")
#draw the median on each box plot
abline(h = median(apply(data2plot, 2, median)),
       col = "green", lwd = 0.6, lty = "dashed")
```

### Density plot

Density plot of the filtered expression dataset before normalization. We can see that the counts are already on a comparable scale, and as stated previously, the normalization may not have a significant impact on the data.

```{r Density plot}
counts_density <- apply(log2(edgeR::cpm(ntm_exp_filtered[,6:92])), 
                        2, density)
  #calculate the limits across all the samples
    xlim <- 0; ylim <- 0
    for (i in 1:length(counts_density)) {
      xlim <- range(c(xlim, counts_density[[i]]$x)); 
      ylim <- range(c(ylim, counts_density[[i]]$y))
    }
    cols <- rainbow(length(counts_density))
    ltys <- rep(1, length(counts_density))
    #plot the first density plot to initialize the plot
    plot(counts_density[[1]], xlim=xlim, ylim=ylim, type="n", 
         ylab="Smoothing density of log2-CPM", 
         main="Density plot before normalization", cex.lab = 0.7)
    #plot each line
    for (i in 1:length(counts_density)) 
      lines(counts_density[[i]], col=cols[i], lty=ltys[i])
    #create legend
    legend("topright", colnames(data2plot),  
           col=cols, lty=ltys, cex=0.45, 
           border ="blue",  text.col = "green4", 
           merge = TRUE, bg = "gray90", ncol = 3)

```

## MA plot comparing the C1-T1 and C1-N1 sample in data set

We can see that most the genes are clustered, but there are some extremes that are spread out. We aim to get rid of these extreme values from our data.
```{r MA plot}
limma::plotMA(log2(ntm_exp[,c(6,7)]), ylab="M - ratio log expression", 
       main="C1-T1 vs C1-N1 MA plot")
```

## Applying TMM
Create an edgeR container for RNASeq count data
```{r results='hide'}
celltypes <-colnames(ntm_exp)[6:92]
filtered_data_matrix <- as.matrix(ntm_exp_filtered[, 6:92])
rownames(filtered_data_matrix) <- ntm_exp_filtered$Gene
d <- edgeR::DGEList(counts = filtered_data_matrix, group = samples$tissue_type)
```

Calculate normalization factor
```{r Normalized factor, results='hide'}
d <- calcNormFactors(d)
```

Get the normalized data
```{r Get normalized data, results='hide'}
#get the normalized data
normalized_counts <- cpm(d)
```



### Boxplot after normalization

The difference between the boxplot before and after normalization is too minimal to be seen, as expected.
```{r Boxplot after normalization, warning=FALSE,message=FALSE}
data3plot <- log2(cpm(normalized_counts[,1:87]))
boxplot(data3plot, xlab = "Samples", ylab = "log2 CPM",
        las = 2, cex = 0.5, cex.lab = 0.5,
        cex.axis = 0.5, main = "Normalized RNASeq Samples")
#draw the median on each box plot
abline(h = median(apply(data3plot, 2, median)),
       col = "green", lwd = 0.6, lty = "dashed")
```




### Density plot after normalization.

There is no difference in the density plot before and after normalization, or the difference is too minimal to be taken into consideration.
```{r Density plot after normalization}
normalized_counts_density <- apply(log2(edgeR::cpm(normalized_counts[,1:87])), 
                        2, density)
  #calculate the limits across all the samples
    xlim <- 0; ylim <- 0
    for (i in 1:length(normalized_counts_density)) {
      xlim <- range(c(xlim, normalized_counts_density[[i]]$x)); 
      ylim <- range(c(ylim, normalized_counts_density[[i]]$y))
    }
    cols <- rainbow(length(normalized_counts_density))
    ltys <- rep(1, length(normalized_counts_density))
    #plot the density plot
    plot(normalized_counts_density[[1]], xlim=xlim, ylim=ylim, type="n", 
         ylab="Smoothing density of log2-CPM", 
         main="Density plot after normalization", cex.lab = 0.7)
    #plot each line
    for (i in 1:length(normalized_counts_density)) 
      lines(normalized_counts_density[[i]], col=cols[i], lty=ltys[i])
    
    #create legend
    legend("topright", colnames(data3plot),  
           col=cols, lty=ltys, cex=0.45, 
           border ="blue",  text.col = "green4", 
           merge = TRUE, bg = "gray90", ncol = 3)

```



## MA plot comparing the C1-T1 and C1-N1 sample after normalization

There is no visible difference in the MA plot before and after normalization. This indicates that the dataset was already at a high quality, also as shown in the box plot and the density plot above.
```{r Normalized MA plot}
limma::plotMA(log2(normalized_counts[,c(1,2)]), ylab="M - ratio log expression", 
       main="Normalized C1-T1 vs C1-N1 MA plot")
```



Inspect the sample separation using a multidimensional scaling plot or MDS plot post normalization.

* Blue: Normal
* Dark green: Mets
* Black: Tumor

From the plot, we can see that the overall separation between each test condition group (T and M) to normal group is good, indicating a good dataset quality.

```{r Visualized post normalized data}
plotMDS(d, cex = 0.5, labels = rownames(samples),
  col = c("darkgreen", "blue")[factor(samples$tissue_type)])
```


# Dispersion
```{r Dispersion}
model_design <- model.matrix(~samples$individual 
                             + samples$tissue_type+0)
d <- edgeR::estimateDisp(d, model_design)
```

Graphing the dispersion-squared BCV.

We can see that genes with low counts have a higher variation, whereas genes with higher counts have a lower variation towards the expected trend. For our dataset, the trend of the count data falls around the common dispersion line in a BCV plot, indicating that the variance of the data is relatively consistent with the expected trend. 

```{r BCV plot}
plotBCV(d,col.tagwise = "black",col.common = "red", main = "Dispersion-squared BCV Plot")
```



# Identifier mapping
Get all availale marts
```{r List all marts}
listMarts()
#If different version is needed
listEnsemblArchives()[1:10,]
```

Connect to the desired mart
```{r Connect to mart}
ensembl <- useMart("ensembl")
```

Get the set of datasets available
```{r Datasets available}
datasets <- listDatasets(ensembl)
knitr::kable(head(datasets),format = "html")
```

Limit to the human datasets available
```{r Limit to human}
knitr::kable(head(datasets[grep(datasets$dataset,
                  pattern = "sapiens"),]),format = "html")
ensembl <- useDataset("hsapiens_gene_ensembl",mart=ensembl)
```
 
## Building a Biomart Query

How many filters are there?
```{r Number of filters}
dim(listFilters(ensembl))
```

What do those filters look like?
```{r Filters}
knitr::kable(listFilters(ensembl)[1:10,1:2], type="html")
biomart_human_filters <- listFilters(ensembl)
```



## Attributes

This dataset stores the genes in various ways, including HGNC symbols, GenBank IDs, and NCBI accessions. To make sure I obtain as many identifiers as possible, I will convert each type separately to HGNC symbols(except for the ones already represented as HGNC symbols).


Use the function __**searchAttributes**__ to find the attributes I am looking for
```{r Search attribute}
kable(searchAttributes(mart = ensembl, 'hgnc') , format="html") %>%
  row_spec(2, background = "yellow")
```

Use the homo sapiens database from Ensembl:
```{r Obtain Homo sapiens database}
httr::set_config(httr::config(ssl_verifypeer = FALSE))
ensembl <- biomaRt::useMart("ensembl", dataset="hsapiens_gene_ensembl")
```


Obtain the genes that are already mapped to HGNC symbols in the original dataset:
```{r Genes mapped to HGNC}
conversion_stash <- "gene_conversion.rds"
if (file.exists(conversion_stash)) {
  gene_conversion <- readRDS(conversion_stash)
} else {
  gene_conversion <- biomaRt::getBM(attributes = "hgnc_symbol",
                                  filters = "hgnc_symbol",
                                  values = ntm_exp_filtered$Gene,
                                  mart = ensembl)
  saveRDS(gene_conversion, conversion_stash)
}

```



Obtain the dataset with genes that are not represented by HGNC symbols
```{r Genes not HGNC}
not_hgnc <- ntm_exp_filtered[-(which(rownames(normalized_counts) %in% gene_conversion$hgnc_symbol)), ]
```
<br>

## 1. Convert Entrez accessions to HGNC symbols

```{r Convert Entrez}
kable(searchAttributes(mart = ensembl, 'entrez') , format="html")

entrez_conversion <- getBM(attributes = c("entrezgene_accession","hgnc_symbol"),
                    filters = "entrezgene_accession",
                    values = not_hgnc$Gene,
                    mart = ensembl)
# Change the row names of the genes matched to HUGO symbols
rownames(normalized_counts)[which(rownames(normalized_counts) %in% entrez_conversion$entrezgene_accession)] <- entrez_conversion$hgnc_symbol
```



## 2. Convert GenBank IDs to HGNC symbols

According to [HGNC website of BioMart](https://www.genenames.org/help/biomart/) help page, the attribute used to map GenBank IDs to HGNC symbols is named as **INSDC**. 

```{r Convert Genbank}
genbank <- not_hgnc[-(which(not_hgnc$Gene %in% entrez_conversion$entrezgene_accession)), ]
knitr::kable(biomart_human_filters[
  grep(biomart_human_filters$description,pattern="INSDC"),],
      format="html")
```

** Can some of our GenBank accession numbers map to it**?

By limiting the filter to homo sapiens, I obtained no results. Since these GenBank accession numbers are not able to map to any alternative names, I have to ignore them from our dataset.
```{r List genbank conversion result}
dim(getBM(attributes = c("protein_id","hgnc_symbol"),
                    filters = "protein_id",
                    values = genbank$Gene,
                    mart = ensembl))

```

## 3. Get rid of the genes (if any) that are mapped to none of the conversion datasets above.


```{r Remove unmapped genes}
mapped_index <- which((rownames(normalized_counts) %in% entrez_conversion$hgnc_symbol) |
                    (rownames(normalized_counts) %in% gene_conversion$hgnc_symbol))


normalized_counts_annot <- normalized_counts[mapped_index, ]
```

How many genes left?
```{r Rest genes}
dim(normalized_counts_annot)
```

**Are there any duplicates?**

Yes, there are 23 duplicated identifiers.
```{r Number of duplicates}
dup_id <- which(duplicated(rownames(normalized_counts_annot)) == TRUE)
length(dup_id)
```

These are the duplicated identifiers.
```{r List duplicates}
rownames(normalized_counts_annot[dup_id, ])
```

**What are they?**

Most of them are non-coding RNAs and pseudogenes, so we can just remove them.
```{r Description}
gene_description <- getBM(attributes = c("description","hgnc_symbol"),
                    filters = "hgnc_symbol",
                    values = rownames(normalized_counts_annot[dup_id, ]),
                    mart = ensembl)

gene_description
```

Get rid of the duplicated rows
```{r Remove duplicates}
normalized_counts_final <- normalized_counts_annot[-dup_id,]
```

The final dataset:
```{r Final dataset}
kable(head(normalized_counts_final), format = "html")
```





# Interpretation

### What are the control and test conditions of the dataset?
The control is the normal thyroid human tissue, and there are two test conditions, which are thyroid cancer primary tumors and metastases tissue types.

### Why is the dataset of interest to you?
I had previous experience with RNAseq analysis on C.elegans data to study the factors that affect resting membrane potential of neurons, and would like to get familiarized with analyzing multi-omics data to improve my skills. This research applied multi-omics profiling on the tissue samples of chromatin accessibility, RNA abundance and protein abundance, which would be a good start for me to learn about cancer.

### Were there expression values that were not unique for specific genes? How did you handle these?
Yes. There are 2 expression values for genes such as CCDC39, DUXAP8, GGT1. I kept them in the dataset rather than filtering them out, because these duplicates are not significant compared to the total sample size.

### Were there expression values that could not be mapped to current HUGO symbols?
Yes. The dataset includes genes that are represented as GenBank accession numbers. By reading through the manual page from [HGNC website of BioMart](https://www.genenames.org/help/biomart/) help page, it indicates that the GenBank accessions can be converted to HGNC symbols using the "INSDC" attribute. However, the __**getBM**__ function from bioMart mapped none of the GenBank accessions in our dataset to HGNC symbols.

### How many outliers were removed?
I removed a total of 11538 outliers with low counts.

### How did you handle replicates?
There are 23 duplicated genes. By checking their descriptions using the __**description**__ attribute, many of these are non-protein coding RNAs and pseudogenes, so I have sufficient confidence to remove them.

### What is the final coverage of your dataset?
The final dataset included 17368 genes with unique identifiers.


# References




