---
title: "Assignment 1"
author: "ChengYue Zhang"
output:
  html_document:
    toc: yes
    toc_depth: '2'
    df_print: paged
  html_notebook:
    toc: yes
    toc_depth: 2
bibliography: A1.bib
csl: "apa-annotated-bibliography.csl"
---

# Download Packages

* [BiocManager](https://CRAN.R-project.org/package=BiocManager) (@morgan2021a)
* [GEOmetadb](https://bioconductor.org/packages/release/bioc/html/GEOmetadb.html) (@mccarthy2012a)
* [edgeR](https://bioconductor.org/packages/release/bioc/html/edgeR.html) (@robinson2010a, @mccarthy2012a, @chen2016a)
```{r message = FALSE, warning=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE)){
  install.packages("BiocManager")}

if (!requireNamespace("GEOmetadb", quietly = TRUE)){
  BiocManager::install("GEOmetadb")}

if (!requireNamespace("ggplot2", quietly = TRUE)){
  install.packages("ggplot2")}

if (!requireNamespace("knitr", quietly = TRUE)){
  install.packages("knitr")}

if (!requireNamespace("RSQLite", quietly = TRUE)){
  install.packages("RSQLite")}

if (!requireNamespace("edgeR", quietly = TRUE)){
  BiocManager::install("edgeR")}

if (!requireNamespace("limma", quietly = TRUE)){
  BiocManager::install("limma")}

if (!requireNamespace("biomaRt", quietly = TRUE)){
  BiocManager::install("biomaRt")}

if (!requireNamespace("GEOquery", quietly = TRUE)){
  BiocManager::install("GEOquery")}

if (!requireNamespace("Biobase", quietly = TRUE)){
  BiocManager::install("Biobase")}

if (!requireNamespace("BioGenerics", quietly = TRUE)){
  BiocManager::install("BioGenerics")}

if (!requireNamespace("odbc", quietly = TRUE)){
  install.packages("odbc")}

if (!requireNamespace("dplyr", quietly = TRUE)){
  install.packages("dplyr")}

if (!requireNamespace("DBI", quietly = TRUE)){
  install.packages("DBI")}

if (!requireNamespace("kableExtra", quietly = TRUE)){
  install.packages("kableExtra")}

```

Load packages
```{r message = FALSE}
library("GEOmetadb")
library("ggplot2")
library("knitr")
library("RSQLite")
library("edgeR")
library("limma")
library("biomaRt")
library("GEOquery")
library("Biobase")
library("BiocGenerics")
library("odbc")
library("dplyr")
library("DBI")
library("kableExtra")
```

# Select an Expression Data Set

## 1. Download the data

Get the GEO description of my dataset GSE162515
```{r message=FALSE}
gse <- getGEO("GSE162515",GSEMatrix=FALSE)
kable(data.frame(head(Meta(gse))), format = "html")
```
<br>
Information about platform
```{r message=FALSE}
current_gpl <- names(GPLList(gse))[1]
current_gpl_info <- Meta(getGEO(current_gpl))

current_gpl_info$title

current_gpl_info$last_update_date

current_gpl_info$organism

```
<br>

Get the expression data
```{r message=FALSE}
sfiles <- getGEOSuppFiles('GSE162515')
```
<br>

There are four supplement files. Here, I will choose [4] which is the RNAseq raw counts file.
```{r echo = TRUE, message = TRUE}
fnames <- rownames(sfiles)
myexp <- read.delim(fnames[4],header=TRUE,
                check.names = FALSE)
fnames
```
<br>

Information about expression data (i.e. gene names, chromosome, start/end position, sequence length) are from column 88 to 92 in the table. Move these to the first column for better visualization and processing.
```{r}
expInfos <- myexp[, seq(88, 92)]

myexp <- cbind(expInfos, myexp[, seq(-88, -92)])
```
<br>
Check the table
```{r}
kable(myexp[1:15,1:15], format = "html")
```



## 2. Assess data quality

How many genes do I have measurements for?
```{r dimension of loaded data}
dim(myexp)
```
<br>
Are any of genes duplicated? and Why?

```{r}
summarized_gene_counts <- sort(table(myexp$Gene),
                               decreasing = TRUE)
```
<br>
Only output those with frequency greater than 1
```{r}
kable(summarized_gene_counts[which(summarized_gene_counts>1)[1:10]], format = 'html')
```
<br>

So yes, there are duplicated genes, but not very significant. The highest frequency of duplicate is only 2. These might be the PCR duplicates that result from sequencing two or more copies of the exact same cDNA fragment. I do not neet to filter them out.

Right now, the column names for the count data are very informative, as shown below:
```{r}
colnames(myexp)[1:15]
```
### Use the GSE series matrix to make the column names informative

Obtain the series matrix file through GEOquery.
```{r message=FALSE}
gseFile <- getGEO('GSE162515',GSEMatrix=TRUE)
gseInfo <- pData(phenoData(gseFile[[1]]))

```
<br>

Filter out the samples that are not RNA-seq.
```{r}
rnaSeq <- gseInfo[grepl("RNA", gseInfo$title), c(1, 6, 8)]
rnaSeq[1:10,]
```
<br>

Process sample names
```{r results='hide'}
samples <- lapply(rnaSeq$title, FUN=function(x){unlist(strsplit(x,
                        split = " "))[2]})
sample_types <- substr(samples,1,nchar(samples)-1)
newColNames <- c(colnames(myexp)[1:5], sample_types)
colnames(myexp) <- newColNames
```
<br>
**Group the data**

There are three groups of samples, where "T" indicates "Thyroid tumor", "N" indicates "Normal thyroid, and "M" indicates "Cervical LymphNode Metastases". 
```{r}
#get the 2 and third token from the column names
samples <- data.frame(lapply(colnames(myexp)[6:92], FUN=function(x){unlist(strsplit(x,
                        split = "\\-"))[c(2,3)]}))
colnames(samples) <- colnames(myexp)[6:92]
rownames(samples) <- c("individual","tissue_type")
samples <- data.frame(t(samples))
```
<br>
From the paper (@sanghi2021a), the authors indicates that there are 27 "N" type samples, 30 "T" type samples and 30 "M" type samples. Check whether I have obtained all of them.
```{r}
sum(grepl("N", samples$tissue_type))
# [1] 27
sum(grepl("T", samples$tissue_type))
# [1] 30
sum(grepl("M", samples$tissue_type))
# [1] 30
```


# Filter out low counts

```{r}
#translate out counts into counts per million using 
#the edgeR package function cpm
cpms <- edgeR::cpm(myexp[,6:92])
rownames(cpms) <- myexp$Gene
# get rid of low counts
keep <- rowSums(cpms >1) >=3
myexp_filtered <- myexp[keep,]
```


<br>
**What does that do to the dataset?**

The size of our dataset gets smaller.
```{r difference before and after filtering on the expression data}
dim(myexp)
dim(myexp_filtered)
```
<br>
**What does that do to the duplicates?**

The duplicates do not change. The highest frequency of duplicate is still 2.
```{r}
summarized_gene_counts_filtered <- sort(table(myexp$Gene),
                               decreasing = TRUE)

kable(summarized_gene_counts_filtered[
  which(summarized_gene_counts_filtered>1)[1:10]],
  format="html")
```



# Normalization
## Distribution of our data
### Boxplot
```{r warning=FALSE,message=FALSE}
data2plot <- log2(cpm(myexp_filtered[,6:92]))
boxplot(data2plot, xlab = "Samples", ylab = "log2 CPM",
        las = 2, cex = 0.5, cex.lab = 0.5,
        cex.axis = 0.5, main = "RNASeq Samples")
#draw the median on each box plot
abline(h = median(apply(data2plot, 2, median)),
       col = "green", lwd = 0.6, lty = "dashed")
```

### Density plot

Density plot of the filtered expression dataset before normalization.The plot is showing a normal distribution, so we will use TMM for normalization.
```{r}
counts_density <- apply(log2(edgeR::cpm(myexp_filtered[,6:92])), 
                        2, density)
  #calculate the limits across all the samples
    xlim <- 0; ylim <- 0
    for (i in 1:length(counts_density)) {
      xlim <- range(c(xlim, counts_density[[i]]$x)); 
      ylim <- range(c(ylim, counts_density[[i]]$y))
    }
    cols <- rainbow(length(counts_density))
    ltys <- rep(1, length(counts_density))
    #plot the first density plot to initialize the plot
    plot(counts_density[[1]], xlim=xlim, ylim=ylim, type="n", 
         ylab="Smoothing density of log2-CPM", 
         main="", cex.lab = 0.7)
    #plot each line
    for (i in 1:length(counts_density)) 
      lines(counts_density[[i]], col=cols[i], lty=ltys[i])
    #create legend
    legend("topright", colnames(data2plot),  
           col=cols, lty=ltys, cex=0.45, 
           border ="blue",  text.col = "green4", 
           merge = TRUE, bg = "gray90", ncol = 3)

```

## MA plot comparing the C1-T1 and C1-N1 sample in data set
```{r}
limma::plotMA(log2(myexp[,c(6,7)]), ylab="M - ratio log expression", 
       main="C1-T1 vs N1 example")
```

## Applying TMM
Create an edgeR container for RNASeq count data
```{r results='hide'}
celltypes <-colnames(myexp)[6:92]
filtered_data_matrix <- as.matrix(myexp_filtered[, 6:92])
rownames(filtered_data_matrix) <- myexp_filtered$Gene
d <- edgeR::DGEList(counts = filtered_data_matrix, group = samples$tissue_type)
```

Calculate normalization factor
```{r results='hide'}
d <- calcNormFactors(d)
```

Get the normalized data
```{r results='hide'}
#get the normalized data
normalized_counts <- cpm(d)
```



Density plot of the expression dataset after normalization.

There is no difference in the density plot before and after normalization, or the difference is too minimal to be visualized.
```{r}
normalized_counts_density <- apply(log2(edgeR::cpm(normalized_counts[,1:87])), 
                        2, density)
  #calculate the limits across all the samples
    xlim <- 0; ylim <- 0
    for (i in 1:length(normalized_counts_density)) {
      xlim <- range(c(xlim, normalized_counts_density[[i]]$x)); 
      ylim <- range(c(ylim, normalized_counts_density[[i]]$y))
    }
    cols <- rainbow(length(normalized_counts_density))
    ltys <- rep(1, length(normalized_counts_density))
    #plot the density plot
    plot(normalized_counts_density[[1]], xlim=xlim, ylim=ylim, type="n", 
         ylab="Smoothing density of log2-CPM", 
         main="", cex.lab = 0.7)
    #plot each line
    for (i in 1:length(normalized_counts_density)) 
      lines(normalized_counts_density[[i]], col=cols[i], lty=ltys[i])
    
    #create legend
    legend("topright", colnames(data2plot),  
           col=cols, lty=ltys, cex=0.45, 
           border ="blue",  text.col = "green4", 
           merge = TRUE, bg = "gray90", ncol = 3)

```




Inspect the sample separation using a multidimensional scaling plot or MDS plot post normalization.

* Blue: Normal
* Dark green: Mets
* Black: Tumor
```{r visualized post normalized data}
plotMDS(d, labels = rownames(samples),
  col = c("darkgreen", "blue")[factor(samples$tissue_type)])

```


# Dispersion
```{r Dispersion}
model_design <- model.matrix(~samples$individual 
                             + samples$tissue_type+0)
d <- edgeR::estimateDisp(d, model_design)
```

Graphing the dispersion-squared BCV
```{r}
plotBCV(d,col.tagwise = "black",col.common = "red",)

```

Create a visual representation of the mean-variance relationship
```{r}
edgeR::plotMeanVar(d, show.raw.vars = TRUE, show.tagwise.vars=TRUE, 
            show.ave.raw.vars = TRUE,  
            NBline=TRUE,
            show.binned.common.disp.vars = TRUE)
```



# Identifier mapping
Get all availale marts
```{r}
listMarts()
#If different version is needed
listEnsemblArchives()[1:10,]
```

Connect to the desired mart
```{r}
ensembl <- useMart("ensembl")
```

Get the set of datasets available
```{r}
datasets <- listDatasets(ensembl)
knitr::kable(head(datasets),format = "html")
```

Limit to the human datasets available
```{r}
knitr::kable(head(datasets[grep(datasets$dataset,
                  pattern = "sapiens"),]),format = "html")
ensembl <- useDataset("hsapiens_gene_ensembl",mart=ensembl)
```
 
## Building a Biomart Query

How many filters are there?
```{r}
dim(listFilters(ensembl))
```

What do those filters look like?
```{r}
knitr::kable(listFilters(ensembl)[1:10,1:2], type="html")
biomart_human_filters <- listFilters(ensembl)
```



## Attributes

This dataset stores the genes in various ways, including HGNC symbols, GenBank IDs, and NCBI accessions. To make sure I obtain as many identifiers as possible, I will convert each type separately to HGNC symbols(except for the ones already represented as HGNC symbols).


Use the function __**searchAttributes**__ to find the attributes I am looking for
```{r}
kable(searchAttributes(mart = ensembl, 'hgnc') , format="html") %>%
  row_spec(2, background = "yellow")
```

Use the homo sapiens database from Ensembl:
```{r}
httr::set_config(httr::config(ssl_verifypeer = FALSE))
ensembl <- biomaRt::useMart("ensembl", dataset="hsapiens_gene_ensembl")
```


Obtain the genes that are already mapped to HGNC symbols in the original dataset:
```{r}
conversion_stash <- "gene_conversion.rds"
if (file.exists(conversion_stash)) {
  gene_conversion <- readRDS(conversion_stash)
} else {
  gene_conversion <- biomaRt::getBM(attributes = "hgnc_symbol",
                                  filters = "hgnc_symbol",
                                  values = myexp_filtered$Gene,
                                  mart = ensembl)
  saveRDS(gene_conversion, conversion_stash)
}

```



Obtain the dataset with genes that are not represented by HGNC symbols
```{r}
not_hgnc <- myexp_filtered[-(which(rownames(normalized_counts) %in% gene_conversion$hgnc_symbol)), ]
```
<br>

## 1. Convert Entrez accessions to HGNC symbols

```{r}
kable(searchAttributes(mart = ensembl, 'entrez') , format="html")

entrez_conversion <- getBM(attributes = c("entrezgene_accession","hgnc_symbol"),
                    filters = "entrezgene_accession",
                    values = not_hgnc$Gene,
                    mart = ensembl)
# Change the row names of the genes matched to HUGO symbols
rownames(normalized_counts)[which(rownames(normalized_counts) %in% entrez_conversion$entrezgene_accession)] <- entrez_conversion$hgnc_symbol
```



## 2. Convert GenBank IDs to HGNC symbols

According to [HGNC website of BioMart](https://www.genenames.org/help/biomart/) help page, the attribute used to map GenBank IDs to HGNC symbols is named as **INSDC**. 

```{r}
genbank <- not_hgnc[-(which(not_hgnc$Gene %in% entrez_conversion$entrezgene_accession)), ]
knitr::kable(biomart_human_filters[
  grep(biomart_human_filters$description,pattern="INSDC"),],
      format="html")
```

** Can some of our GenBank accession numbers map to it**?

By limiting the filter to homo sapiens, I obtained no results. Since these GenBank accession numbers are not able to map to any alternative names, I have to ignore them from our dataset.
```{r}
dim(getBM(attributes = c("protein_id","hgnc_symbol"),
                    filters = "protein_id",
                    values = genbank$Gene,
                    mart = ensembl))

```

## 3. Get rid of the genes (if any) that are mapped to none of the conversion datasets above.


```{r}
mapped_index <- which((rownames(normalized_counts) %in% entrez_conversion$hgnc_symbol) |
                    (rownames(normalized_counts) %in% gene_conversion$hgnc_symbol))


normalized_counts_annot <- normalized_counts[mapped_index, ]
```

How many genes left?
```{r}
dim(normalized_counts_annot)
```

**Are there any duplicates?**

Yes, there are 23 duplicated identifiers.
```{r}
dup_id <- which(duplicated(rownames(normalized_counts_annot)) == TRUE)
length(dup_id)
```

These are the duplicated identifiers.
```{r}
rownames(normalized_counts_annot[dup_id, ])
```

**What are they?**
```{r}
gene_description <- getBM(attributes = c("description","hgnc_symbol"),
                    filters = "hgnc_symbol",
                    values = rownames(normalized_counts_annot[dup_id, ]),
                    mart = ensembl)


```

Get rid of the duplicated rows
```{r}
normalized_counts_final <- normalized_counts_annot[-dup_id,]
```

The final dataset:
```{r}
kable(head(normalized_counts_final), format = "html")
```





# Interpretation

### What are the control and test conditions of the dataset?
The control is the normal thyroid human tissue, and there are two test conditions, which are thyroid cancer primary tumors and metastases tissue types.

### Why is the dataset of interest to you?
I had previous experience with RNAseq analysis on C.elegans data to study the factors that affect resting membrane potential of neurons, and would like to get familiarized with analyzing multi-omics data to improve my skills. This research applied multi-omics profiling on the tissue samples of chromatin accessibility, RNA abundance and protein abundance, which would be a good start for me to learn about cancer.

### Were there expression values that were not unique for specific genes? How did you handle these?
Yes. There are 2 expression values for genes such as CCDC39, DUXAP8, GGT1. I kept them in the dataset rather than filtering them out, because these duplicates are not significant compared to the total sample size.

### Were there expression values that could not be mapped to current HUGO symbols?
Yes. The dataset includes genes that are represented as GenBank accession numbers. By reading through the manual page from [HGNC website of BioMart](https://www.genenames.org/help/biomart/) help page, it indicates that the GenBank accessions can be converted to HGNC symbols using the "INSDC" attribute. However, the __**getBM**__ function from bioMart mapped none of the GenBank accessions in our dataset to HGNC symbols.

### How many outliers were removed?
I removed a total of 11538 outliers with low counts.

### How did you handle replicates?
There are 23 duplicated genes. By checking their descriptions using the __**description*__ attribute, many of these are non-protein coding RNAs and pseudogenes, so I have sufficient confidence to remove them.

### What is the final coverage of your dataset?
The final dataset included 17345 genes with unique identifiers.


# References




